Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/ApayaTokenizationViewModel.swift

```diff
@@ -423,14 +405,33 @@ extension ApayaTokenizationViewModel: WKNavigationDelegate {
     
 }
 
+extension ApayaTokenizationViewModel {
+        
+    private func selectPaymentMethodWithParametersIfNeeded(_ parameters: [String: Any]) {
+        
+        firstly {
+            ClientSession.Action.selectPaymentMethodWithParametersIfNeeded(parameters)
+        }
+        .done {}
+        .catch { error in
+            self.handleErrorBasedOnSDKSettings(error)
+        }
+    }
+}
+
 extension ApayaTokenizationViewModel {
     
     override func handle(error: Error) {
-        ClientSession.Action.unselectPaymentMethod(resumeHandler: nil)
-        self.completion?(nil, error)
-        self.completion = nil
+        firstly {
+            ClientSession.Action.unselectPaymentMethodIfNeeded()
+        }
+        .ensure {
+            self.executeCompletionAndNullifyAfter(error: error)
```

Comment:
Problem with this is that the merchant is sending an error (his error) that you should handle, I don't think he should wait until we make an API call to show their error.

Second problem is that you're calling `executeCompletionAndNullifyAfter` which will propagate the error, and call `self.handleFailedTokenizationFlow(error: err)` from the `startTokenizationFlow`, while you're also calling it on the line below.